main/src/ca/uwaterloo/flix/language/phase/unification/set/SetUnification.scala

```diff
@@ -0,0 +1,235 @@
+
+        // f1 ∪ f2 ∪ ... ~ empty
+        // ---
+        // [f1 ~ empty, f2 ~ empty, ...],
+        // {}
+        case (union@Union(_, _, _, _, _, _, _), Empty) =>
+          p.progress()
+          val eqs = union.mapSubformulas(Equation.mk(_, Empty, loc))
+          (eqs, SetSubstitution.empty)
+
+        // f1 ~ f2, where f1 and f2 are ground
+        // ---
+        // [], {} if solved
+        // [f1 !~ f2], {} if unsolvable
+        case (f1, f2) if f1.isGround && f2.isGround =>
+          p.progress()
+          if (isEquivalent(f1, f2)) (Nil, SetSubstitution.empty)
+          else (List(eq.toUnsolvable), SetSubstitution.empty)
+
+        case _ =>
+          // Can't do anything.
+          (List(eq), SetSubstitution.empty)
+      }
+    }
+
+    /**
+      * Solves equations variable alias equations (e.g. `x1 ~ x2`).
+      *
+      *   - `x1 ~ x2` becomes `({}, [x1 -> x2])`
+      *   - `!x1 ~ !x2` becomes `({}, [x1 -> x2])`
+      */
+    def variableAlias(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(t1, t2, _, _) = eq
+      (t1, t2) match {
+        // x1 ~ x2
+        // ---
+        // {},
+        // [x1 -> x2]
+        case (Var(x), y@Var(_)) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, y))
+
+        // !x1 ~ !x2
+        // ---
+        // {},
+        // [x1 -> x2]
+        case (Compl(Var(x)), Compl(y@Var(_))) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, y))
+
+        case _ =>
+          // Can't do anything.
+          (List(eq), SetSubstitution.empty)
+      }
+    }
+
+    /**
+      * Solves non-recursive variable assignments (e.g. `x1 ~ x2 ∪ c4`).
+      *
+      *   - `x ~ f` where `f` does not contain `x` becomes `([], {x -> f})`
+      *   - `!x ~ f` where `f` does not contain `x` becomes `([], {x -> !f})`
+      */
+    def variableAssignment(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
+      val Equation(t1, t2, _, _) = eq
+      (t1, t2) match {
+        // x ~ f, where f does not contain x
+        // ---
+        // {},
+        // [x -> f]
+        case (v@Var(x), f) if !f.contains(v) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, f))
+
+        // !x ~ f, where f does not contain x
+        // ---
+        // {},
+        // [x -> !f]
+        case (Compl(v@Var(x)), f) if !f.contains(v) =>
+          p.progress()
+          (Nil, SetSubstitution.singleton(x, mkCompl(f)))
+
+        case _ =>
+          // Can't do anything.
+          (List(eq), SetSubstitution.empty)
+      }
+    }
+
+    /**
+      * Solves equations using successive-variable-elimination, i.e. exhaustive instantiation.
+      *
+      * Always returns no equations or `eq` marked as [[Equation.Status.Unsolvable]] or
+      * [[Equation.Status.Timeout]].
+      */
+    def sve(recSizeThreshold: Int)(eq: Equation)(implicit p: Progress): (List[Equation], SetSubstitution) = {
```

Comment:
Any reason recSizeThreshold is an argument? Could it not be a static final field, so to speak?

Alternatively, it could be a property on `Progress` which could potentially be renamed to `Solver`?

Reply:
To have it be changable - i was annoyed by the static thresholds of the old solver

Hmm we could make `Progress` the "settings" object but I want to make the rules as selfcontained and simple as possible - removed from the other infrastructure. So I think direct arguments and simple objects achieve that better than syntactic brevity

Reply:
Changeable by who?

Reply:
without static mutation, by anyone. If we want it as a compiler flag, need to disable it for testing, etc